src/System.IO.Pipelines/ReadableBufferReader.cs

```diff
@@ -53,78 +54,107 @@ public ReadCursor Cursor
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Peek()
         {
-            if (_end)
+            if (!_end)
             {
-                return -1;
+                return _currentSpan[_index];
             }
-            return _currentSpan[_index];
+
+            return -1;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Take()
         {
-            if (_end)
+            if (!_end)
```

Comment:
Have to reverse the truth/false to get it to create right codegen; adding commit